### PR TITLE
Sema: Fix compiler crash when resolving Extension declaration with UnboundGenericType

### LINKED ISSUE #60922: lib/AST/RequirementMachine/Symbol.cpp:81: Assertion `!type->hasUnboundGenericType()' failed
**Describe the bug**

the following (invalid) test program crashes the compiler:

```swift 
enum Selection<Element>
{
    case one(Element)
    case many([Element])
}
extension Selection?
{
}
```

```
$ ~/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-08-30-a/usr/bin/swiftc test.swift
error: compile command failed due to signal 6 (use -v to see invocation)
swift-frontend: /home/build-user/swift/lib/AST/RequirementMachine/Symbol.cpp:81: swift::rewriting::Symbol::Storage::Storage(Symbol::Kind, swift::CanType, ArrayRef<swift::rewriting::Term>): Assertion `!type->hasUnboundGenericType()' failed.
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the project and the crash backtrace.
Stack dump:
0.      Program arguments: /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-08-30-a/usr/bin/swift-frontend -frontend -c -primary-file test.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -color-diagnostics -new-driver-path /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-08-30-a/usr/bin/swift-driver -empty-abi-descriptor -resource-dir /home/ec2-user/.swiftenv/versions/DEVELOPMENT-SNAPSHOT-2022-08-30-a/usr/lib/swift -module-name test -o /tmp/TemporaryDirectory.GFWOdM/test-1.o
1.      Swift version 5.8-dev (LLVM 57827dcce6fa0fb, Swift 04f1cde82e1d1f3)
2.      Compiling with the current language version
3.      While evaluating request TypeCheckSourceFileRequest(source_file "test.swift")
4.      While type-checking extension of Selection? (at test.swift:6:1)
5.      While evaluating request GenericSignatureRequest(extension of Selection?)
6.      While evaluating request InferredGenericSignatureRequest(<Wrapped>, (null), extension of Selection?, {Wrapped == Selection}, {}, 1)
```

**Steps To Reproduce**
1. `swiftc` the given test program, on a recent nightly (i reproduced the crash on `DEVELOPMENT-SNAPSHOT-2022-08-18-a` and `DEVELOPMENT-SNAPSHOT-2022-08-30-a`.)

**Expected behavior**

the compiler should not crash

**Environment (please fill out the following information)**
 - OS: amazon linux 2
- swift: `DEVELOPMENT-SNAPSHOT-2022-08-30-a`

### PR DESCRIPTION
**Pull request by simanerush**
Resolves #60922.

### CODE REVIEW

File: lib/Sema/TypeChecker.h [hunk 1]
```diff
@@ -924,7 +924,8 @@ enum class UnsupportedMemberTypeAccessKind : uint8_t {
 /// member of the given base type.
 UnsupportedMemberTypeAccessKind
 isUnsupportedMemberTypeAccess(Type type, TypeDecl *typeDecl,
-                              bool hasUnboundOpener);
+                              bool hasUnboundOpener,
+                              bool isExtensionBinding = false);
```
Comment (slavapestov) [inline lib/Sema/TypeChecker.h:928]: Instead of passing two booleans, maybe it's better to pass the TypeResolverOptions?

Comment (simanerush) [reply lib/Sema/TypeChecker.h:928]: Do you mean a tuple of booleans?

Comment (simanerush) [reply lib/Sema/TypeChecker.h:928]: This is a good idea; however, in `resolveNestedIdentTypeComponent` function, we are retrieving `hasUnboundOpener`boolean value from `TypeResolution`, not from its options. So I don't think passing options here would b enough. As an alternative, though, I can think of a way of somehow combining these two flags into one.

File: test/Generics/unbound.swift [hunk 2]
```diff
@@ -137,3 +137,19 @@ do {
   assertExactType(of: x1, is: Generic<Int, Bool>.self)
   assertExactType(of: x2, is: Generic<Int, Bool>.self)
 }
+
+// https://github.com/apple/swift/issues/60922
+enum T<E> {
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:142]: Could you transpose the names of the enum and generic parameter to make it feel more natural?

File: test/Generics/unbound.swift [hunk 3]
```diff
@@ -137,3 +137,19 @@ do {
...
+  // expected-note@-1{{generic type 'T' declared here}}
+  case a(E)
+  case b([E])
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:145]: The cases can me omitted; they aren’t involved anywhere in these tests or the code paths being tested.

File: test/Generics/unbound.swift (hunk 2 above)
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:142]: Needs a blank line.

File: test/Generics/unbound.swift [hunk 4]
```diff
@@ -137,3 +137,19 @@ do {
...
+}
+
+extension T? {}
+// expected-error@-1{{reference to generic type 'T' requires arguments in <...>}}
+
+struct G<A, B>: P {}
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:151]: Why the conformance?

File: test/Generics/unbound.swift [hunk 5]
```diff
@@ -137,3 +137,19 @@ do {
...
+// expected-note@-1{{generic type 'G' declared here}}
+
+extension G? {}
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:154]: Could you also add an extension of the desugared version of either of these types?

File: test/Generics/unbound.swift (hunk 4 above)
Comment (simanerush) [reply test/Generics/unbound.swift:151]: Not needed here. Will fix that

File: test/Generics/unbound.swift [hunk 6]
```diff
@@ -137,3 +137,23 @@ do {
   assertExactType(of: x1, is: Generic<Int, Bool>.self)
   assertExactType(of: x2, is: Generic<Int, Bool>.self)
 }
+
+// https://github.com/apple/swift/issues/60922
+
+enum E<T> {}
+// expected-note@-1{{generic type 'E' declared here}}
+// expected-note@-2{{generic type 'E' declared here}}
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:145]:
```suggestion
// expected-note@-1 2 {{generic type 'E' declared here}}
```

File: test/Generics/unbound.swift [hunk 7]
```diff
@@ -137,3 +137,23 @@ do {
...
+extension Optional<E> {}
+// expected-error@-1{{reference to generic type 'E' requires arguments in <...>}}
+
+struct G<T> {}
+// expected-note@-1{{generic type 'G' declared here}}
+// expected-note@-2{{generic type 'G' declared here}}
+
+extension G? {}
+// expected-error@-1{{reference to generic type 'G' requires arguments in <...>}}
+extension Optional<G> {}
```
Comment (AnthonyLatsis) [inline test/Generics/unbound.swift:158]: Just one "desugared" extension is enough, I believe.

**Review by slavapestov — APPROVED** [review]

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@simanerush You can go ahead and merge this if you're ready.

**Comment by simanerush** [comment]
@slavapestov Unfortunately, I don't have write access to merge this. Maybe you or @AnthonyLatsis could do that

**Comment by xedin** [comment]
@swift-ci please smoke test Windows platform

**Pull request merged**